### PR TITLE
[c-interop][reverse][kt-36878] unifiyng generation of type handlers inlined and not inlined types

### PR DESCRIPTION
**Pull request by vvlevchenko**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CAdapterGenerator.kt [hunk 1]
```diff
@@ -797,20 +798,14 @@ internal class CAdapterGenerator(val context: Context) : DeclarationDescriptorVi
...
         val set = mutableSetOf<ClassDescriptor>()
         defineUsedTypesImpl(scope, set)
-        // Add nullable primitives.
-        predefinedTypes.forEach {
-            val nullableIt = it.makeNullable()
+        val types = predefinedTypes + set
+                .map {  if (it.isInline) it.underlyingRepresentation()!!.type else it.defaultType }
+                .filter{isMappedToReference(it)}
+        val nullableTypes = types.map { it.makeNullable() }.toSet()
+        nullableTypes.forEach {
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CAdapterGenerator.kt:805]: Does this work for 
```kotlin
inline class Foo(val value: Any?)
inline class Bar(val value: Foo)

fun takeNullableBar(nullableBar: Bar?) {}
```
?

Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CAdapterGenerator.kt:804]: I've added support inline and recursive inline classes for arguments and return types.

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CAdapterGenerator.kt:804]: Please add the code above to the test suite.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CAdapterGenerator.kt [hunk 2]
```diff
@@ -297,12 +296,12 @@ private class ExportedElement(val kind: ElementKind,
             original is ConstructorDescriptor ->
                 SignatureElement(uniqueName(original, shortName), original.constructedClass.defaultType)
             else ->
-                SignatureElement(uniqueName(original, shortName), original.returnType!!)
+                SignatureElement(uniqueName(original, shortName), original.returnType!!.noInlineUnderlyingRepresentation() ?: original.returnType!!)
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CAdapterGenerator.kt:299]: What is the purpose of this patch? Does it change how types are mapped?

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CAdapterGenerator.kt:299]: I see.
New mapping behaviour seems incorrect. For `takeNullableBar` from [the motivational test case](https://github.com/JetBrains/kotlin-native/pull/4200/commits/ac0a57c3f7df6cc3d7834bad2ade56ae5807d701#r434619545) it should produce `Bar` as before, not `Any`.

Mapping was ok, it is generation of used types that was broken.

Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CAdapterGenerator.kt:299]: Hm, I'm totally confused now. current (master) behaviour is: avoid type generation for inline classes. and my patch prevent appearing "unwanted" types in return and as parameter. Or I incorrectly understand the idea?

**Comment by SvyatoslavScherbina** [comment]
To be moved to https://github.com/JetBrains/kotlin.

**Pull request closed without merging**